### PR TITLE
cleanup of UDP behavior

### PR DESCRIPTION
**Pull request by doudou**
This is built on top of #33. Only the last commit is relevant.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -129,6 +129,40 @@ Examples:
...
+
+For backward compatibility reasons, the default values for `connected` and
+`ignore_connrefused` depends on whether the UDP stream was created with or without
+a local port. The behavior described below is deprecated. In the future, the default
+will be set to `connected=1` and `ignore_connrefused=1`. Warnings are currently issued
+when the current defaults are used. Explicitely set these parameters to shut the
+warnings and ensure your code will continue working as-is when the defaults change.
+
+- `ignore_connrefused` may be set to zero (to report connrefused) only on connected
+  sockets. Attempting to set `ignore_connrefused=0&connected=0` will throw in `openURI`
```
Comment (g-arjones) [inline README.md:155]: Is there a reason for not allowing to ignore ECONNREFUSED on unconnected sockets?

File: src/IOStream.cpp [hunk 2]
```diff
@@ -160,11 +169,15 @@ size_t UDPServerStream::write(uint8_t const* buffer, size_t buffer_size)
...
 
     ssize_t ret = sendto(m_fd, buffer, buffer_size, 0, &m_si_other, m_s_len);
-    if (ret == -1 && errno != EAGAIN && errno != ENOBUFS){
+    if (ret == -1) {
+        if (errno == EAGAIN && errno == ENOBUFS) {
+            return 0;
+        }
+        else if (m_ignore_econnrefused && errno == ECONNREFUSED) {
+            return 0;
+        }
```
Comment (g-arjones) [inline src/IOStream.cpp:178]: You don't get this on send()/sendto(). This is an ICMP response you receive from the remote host so you get it on the subsequent recv() call

File: README.md (hunk 1 above)
Comment (doudou) [reply README.md:155]: According to my test, recvfrom does not report ECONNREFUSED on non-connected sockets.

File: src/IOStream.cpp (hunk 2 above)
Comment (doudou) [reply src/IOStream.cpp:178]: I thought the same, but it's not true. You get it on the first recvfrom/sendto that interacts with the socket after the reception of the ICMP message.

File: README.md [hunk 3]
```diff
@@ -129,6 +129,40 @@ Examples:
 - `udp://localhost:4000`
 - `udp://localhost:4000?local_port=4001`
 
+**The Connection Refused error** If configured to do so, UDP streams will report
+a connection refused error if there are no processes listening on the configured
```
[the remark this answers is not in the document]
Comment (doudou) [reply README.md:133]: Cf answer to the other comment: it seems that we get it on the first recvfrom/sendto that interacts with the socket after the reception of the ICMP message. That's why I added the code to check it, as well as the corresponding tests.

I will definitely add the description of the behavior to the documentation.

File: src/IOStream.cpp (hunk 2 above)
Comment (g-arjones) [reply src/IOStream.cpp:178]: Interesting. That's not what linux manpages says. Anyways... :+1:

File: README.md (hunk 1 above)
Comment (g-arjones) [reply README.md:155]: Didn't know that. :+1:

File: src/IOStream.cpp (hunk 2 above)
Comment (doudou) [reply src/IOStream.cpp:178]: I have to admit that I had the same understanding than you ... but then you piqued my interest

From udp(7): All errors documented for socket(7) or ip(7) may be returned by a send **or receive** on a UDP socket.

Then lists ECONNREFUSED as a special udp-specific error.

Then, sendto: "Additional errors may be generated and returned from the underlying protocol modules; see their respective manual pages."

It seems to be there. It's just very hard to spot.

Comment (g-arjones) [reply src/IOStream.cpp:178]: Ah, there it is. I missed the "additional errors may be generated (,,,)" part in sendto(). Thanks for clarifying...

**Comment by doudou** [comment]
Updated

**Comment by g-arjones** [comment]
Great!

**Pull request merged**